### ISSUE TITLE
Get organisation for some users returns empty result (ldap+auth.poxy config)

### ISSUE THREAD

**Issue by tsn77130** [grafana/grafana#11311]
Hi all,

When I login with new user, I fall into -1 org (auth.proxy + LDAP config) on 1st log.
Problem is that I sometimes can't get out of this org, because some of my users does not seems to have another org declared in grafana, despite fact that LDAP & ldap.toml configuration are OK.
That seems to be random problem, some users can change org, and other can't.

Example : 

```
#User exists
curl -s -XGET "http://127.0.0.1:3000/api/users/lookup?loginOrEmail=xxx@xxx.com" -u "xxx:yyy"
{
   "isGrafanaAdmin" : false,
   "name" : " 7164",
   "email" : "xxx@xxx.com",
   "theme" : "",
   "id" : 113,
   "orgId" : -1,
   "login" : "xxx@xxx.com"
}

As you can see, user falls into -1 org ...

#Org exists
curl -s -XGET "http://127.0.0.1:3000/api/orgs/7164" -u "xxx:yyy" | json_pp   
{
   "address" : {
      "city" : "",
      "address1" : "",
      "state" : "",
      "country" : "",
      "address2" : "",
      "zipCode" : ""
   },
   "id" : 7164,
   "name" : "7164"
}

#If I check user orgs, I see this is empty, it should display user's org "7164" according to ldap config.
curl -s -XGET "http://127.0.0.1:3000/api/users/113/orgs" -u "xxx:yyy"
[]

#With some other users, I have correct group
curl -s -XGET "http://127.0.0.1:3000/api/users/97/orgs" -u "xxx:yyy"
[{"orgId":7023,"name":"7023","role":"Viewer"}]

#So if I try to change user org ID, this does not works (that seems logic)
curl -XPOST "http://127.0.0.1:3000/api/users/113/using/7164" -u "xxx:yyy"
{
   "message" : "Not a valid organization"
}

#ldap.toml group mapping
[[servers.group_mappings]]
group_dn = "cn=7164,ou=CustomersGroups,dc=xxx,dc=com"
org_role = "Viewer"
org_id = 7164


#And LDAP search which is return correct result (same group defined in ldap.toml)
ldapsearch -x -D cn=admin,dc=xxx,dc=com -w "xxx:yyy" -b ou=CustomersGroups,dc=xxx,dc=com AccountID=7164 member 

 7164, CustomersGroups, xxx.com
dn: cn=7164,ou=CustomersGroups,dc=xxx,dc=com
member: uid=xxx@xxx.com,ou=Customers,dc=xxx,dc=com

 search result
search: 2
result: 0 Success

 numResponses: 2
 numEntries: 1

```

If I do the exact same procedure with another account, it sometimes works, and sometimes not.
LDAP search logs are exactly the same between one account which works, and another one which does not.

Is there something which could explain this ?

What Grafana version are you using? 4.6.4 & 5.0.1 tested
What datasource are you using? influxdb
What OS are you running grafana on? debian 9 
What did you do? I tried to login with user defined in LDAP group mapping
What was the expected result? Login, ideally directly in correct Org, at least in -1 org but with possibility to switch to correct org.
What happened instead? I can't switch to user org
 
Thanks, thomas

**Comment by svet-b** [comment]
Just wanted to confirm I've experienced similar glitchy-ness with respect to users ending up logged into org -1, with a similar setup. I've never quite managed to reliably reproduce the issue or get to the bottom of it...logging out and logging back in generally seems to fix it. Would be great if the root cause could be identified.

**Comment by tsn77130** [comment]
I thought I had found a clue, with limit 1000 in func SearchOrgs in org.go, because some orgs wasn't displayed (only the first 1000), but after org cleaning , these orgs now displays but my users are still not associated with.
when I login with new user, I still get -1 org, and still can't switch to another org, because user seems to have no other org (despite ldap configuration is OK)
Curl returns empty result.
When I try to force user's org (with admin UI login, or API request), it seems good (I can add org to user), but when i try to login I still have no orgs, and when I do curl again, user/xx/org is still empty...

**Comment by rda0** [comment]
We are experiencing the same issues, reproducible.

- 1) User mapping to one org + org_role only gets `/?orgId=-1`
- 2) Removing User from LDAP group does not remove grafana permissions/roles

# Steps  to reproduce 1):

If a user maps via `ldap.toml` to one org only, after logging in the following URL is shown in the browser:

    https://<our_domain>/?orgId=-1

For example when the user maps to `orgId=2` with `Editor` permissions, no dashboards are shown, nor to add dashboards, nor to change/show the users org. A one time manual change of the URL in the browser to:

    https://<our_domain>/?orgId=2

fixes it and shows the dashboards. After logging out and logging back in, the URL is still the same (`?orgId=2`) nd it works.

This does not happen if the user also maps to the `org=1`. It seems like this has something to do with the option:

    [users]
    # Set to true to automatically assign new users to the default organization (id 1)
    auto_assign_org = true

Although if I set this to `false` LDAP user login does not work at all, showing the following error message to the user in the browser:

    Error while trying to authenticate user

and the following error message in the logs:

    lvl=eror msg="Error while trying to authenticate user" logger=context userId=0 orgId=0 uname= error="Cannot remove last organization admin"

# Steps  to reproduce 2):

If this user is now removed from this LDAP group, which maps to to `orgId=2` with `Editor` permissions. The user can still login and has the same permissions as before. Even if I login as a "Grafana Admin", remove the role in org 2 for the respective user and then remove the user itself, the user is still able to login using the same permissions (`Editor`, `org=2`).

# Versions

What Grafana version are you using? `5.0.3`
What datasource are you using? `influxdb`
What OS are you running grafana on? `debian 9`

# Configuration

## grafana.ini

```
[paths]
[server]
protocol = http
http_addr = 127.0.0.1
http_port = 3000
domain = <our_domain>
root_url = https://%(domain)s/
enable_gzip = true
[database]
log_queries =
[session]
[dataproxy]
[analytics]
[security]
admin_user = <our_admin_user>
admin_password = <our_admin_pw>
secret_key = <our_secret_key>
[snapshots]
[dashboards]
[users]
allow_sign_up = false
allow_org_create = false
auto_assign_org = true
login_hint = dphys username
[auth]
[auth.anonymous]
[auth.github]
[auth.google]
[auth.generic_oauth]
[auth.grafana_com]
[auth.proxy]
[auth.basic]
[auth.ldap]
enabled = true
allow_sign_up = true
[smtp]
[emails]
[log]
[log.console]
[log.file]
[log.syslog]
[alerting]
[metrics]
[metrics.graphite]
[tracing.jaeger]
[grafana_com]
[external_image_storage]
[external_image_storage.s3]
[external_image_storage.webdav]
[external_image_storage.gcs]
[external_image_storage.azure_blob]
[external_image_storage.local]
```

## ldap.toml

```
[[servers]]
host = "ldap1... ldap2... ldap3..."
port = 636
use_ssl = true
start_tls = false
ssl_skip_verify = false
bind_dn = "uid=%s,ou=people,<our_base>"
search_filter = "(uid=%s)"
search_base_dns = ["<our_base>"]
group_search_filter = "(&(objectClass=posixGroup)(memberUid=%s))"
group_search_base_dns = ["ou=groups,<our_base>"]
[servers.attributes]
name = "givenName"
surname = "sn"
username = "uid"
member_of = "cn"
email =  "mail"
[[servers.group_mappings]]
group_dn = "admingroup"
org_role = "Admin"
org_id = 1
[[servers.group_mappings]]
group_dn = "testgrafana1"
org_role = "Admin"
org_id = 2
[[servers.group_mappings]]
group_dn = "testgrafana2"
org_role = "Editor"
org_id = 2
[[servers.group_mappings]]
group_dn = "testgrafana3"
org_role = "Viewer"
org_id = 2
```

**Comment by tsn77130** [comment]
@rda0 , for step 1, I had same error message. I fixed by deleting by UI all global users (with admin account). And I still have auto_assign_org = false)

After 1st login, I was in orgId=-1, with no button for switching. Changing URL wasn't working for me (page not found)
I fixed it by using https://<our_domain>/profile/switch-org/'org_id' after 1st login (and I used a redirection trick to always call this page when users login now, it is idempotent.)

It does not work as it should be, but with these fixs, it is now usable for us.

**Comment by rda0** [comment]
@tsn77130, thank you.

I now use another workaround, using `auto_assign_org = true` and my first `ldap.toml` mappings are:

```
[[servers.group_mappings]]
group_dn = "admingroup"
org_role = "Admin"
org_id = 1

[[servers.group_mappings]]
group_dn = "*"
org_role = "Viewer"
org_id = 1
```

I use the `orgId=1` as group for all our users with viewer permissions. The dashboard will only contain a description, but nothing more. This fixes the issue when logging in for the first time.

But  when a user was active in another org (i.e.: `orgId=3`) and removed from the respective ldap group. At the next login he will be in `orgId=-1`. I use this rewrite rule (apache), but I have to refresh the page once for it to take effect:

```
RewriteEngine On
RewriteCond "%{QUERY_STRING}" "^orgId=-1$"
RewriteRule "(.*)" "?orgId=1" [R=temp,L]
```

Do you have a better solution?

**Comment by DanCech** [comment]
#11354 changes the way that user/org role mappings are managed by the LDAP integration, and should solve this problem.  Can you verify whether the issue is fixed in releases >=5.1?

**Comment by rda0** [comment]
@DanCech I confirm the issue is fixed for us. Version 5.1.3
